artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/transaction/ProtonTransactionHandler.java

```diff
@@ -165,6 +170,16 @@ public void onError(int errorCode, String errorMessage) {
       }
    }
 
+   private void cancelRefs(ProtonTransactionImpl tx) throws Exception {
+      List<MessageReference> toCancel = new ArrayList<>();
```

Comment:
The only bit i wonder if we need to add here is the extra logic of stopping and restarting the consumer as like in ServerSessionImpl::cancelAndRollback

Reply:
actually will be cleaner just to invoke rollback in serversession impl, then will have same logic, and will gain from having any other fixes.